[2단계 - 웹 자동차 경주] 민트(유재민) 미션 제출합니다.

yujamint opened the pull request:
안녕하세요 완태! 리뷰이 민트입니다.
2단계 구현 완료하여 리뷰 요청드립니다.
1단계에서 좋은 리뷰 해주셔서 감사합니다. 필요한 지식에 대해 궁금해하고, 고민해볼 수 있는 좋은 기회였습니다 :)
2단계 진행하며 궁금했던 내용은 코멘트를 통해 질문 남기겠습니다. 답변해주시면 감사하겠습니다 😄
이번 리뷰도 잘 부탁드리겠습니다 완태!!

+) 완태가 스프링에서의 예외 처리에 대해 작성한 테코블 잘 읽었습니다!! ㅎㅎ

yujamint committed:
refactor: RestController 어노테이션 사용하도록 수정

yujamint committed:
refactor: 불필요한 Autowired 제거

yujamint committed:
refactor: dto 패키지 이동

yujamint committed:
docs: 리팩토링 목록 작성

yujamint committed:
docs: 2단계 요구사항 목록 작성

yujamint committed:
refactor: ResponseDto 클래스명 변경

게임 실행에 대한 응답임을 의미하도록 수정

yujamint committed:
refactor: sql 파일 이름 변경

테이블의 스키마를 정하는 쿼리문이기 때문에 schema.sql로 변경

yujamint committed:
feat: 조인 쿼리 테스트를 위한 초기 데이터 쿼리 추가

yujamint committed:
refactor: GamePlayResponseDto 생성자 수정

yujamint committed:
docs: 조인 쿼리 테스트용 데이터 추가

yujamint committed:
refactor: winners만 입력받는 DTO 생성자 추가

yujamint committed:
feat: 게임 플레이 이력 조회 API 구현

yujamint committed:
docs: 게임 플레이 이력 조회 API 구현 - 완료 처리

yujamint committed:
refactor: 게임 결과 출력 뷰 API 반환 타입에 맞게 수정

yujamint committed:
chore: 사용하지 않는 코드 임시 주석 처리

yujamint committed:
feat: 콘솔 애플리케이션에서도 플레이 결과 출력하도록 구현

yujamint committed:
refactor: Location 클래스명 Position으로 수정

yujamint committed:
refactor: 웹 애플리케이션으로 수정하며 불필요해진 코드 제거

yujamint committed:
refactor: 잘못된 입력값에 대해 예외 던지도록 수정

- 게임 시도 횟수가 1 미만일 때
- 중복되는 차 이름을 입력했을 떄

yujamint committed:
feat: 서버에서 보낸 에러 메시지 alert 띄우는 기능 구현

yujamint committed:
feat: 예외 발생 시 response로 전송할 정보 객체 구현

yujamint committed:
feat: ControllerAdvice를 사용한 예외 처리 기능 구현

IllegalArgumentException 처리

yujamint committed:
refactor: Names 일급 컬렉션 생성을 통한이름 중복 검사 로직 분리

yujamint committed:
refactor: Name 유효성 검사 로직 가독성 개선

- 메서드명 변경
- Names에서 trim을 하므로, 불필요한 검사 코드 제거

yujamint committed:
chore: data-test.sql 파일이 API 테스트 데이터임을 알리는 주석문 추가

yujamint committed:
docs: 2단계 요구사항 완료 처리

yujamint commented on src/main/java/racingcar/dao/InMemoryGameDao.java:
```diff
@@ -31,4 +36,36 @@ public int save(int trialCount, String winners) {
     }
+
+    @Override
+    public List<GamePlayResponseDto> selectAll() {
```
### 두 테이블을 JOIN하며 조회하는 책임은 어떤 DAO가 갖는 게 좋을까요?
요구사항에 따라 이력 조회 API를 구현하며 CAR_RESULT 테이블과 PLAY_RESULT 테이블을 JOIN하며 조회를 해야 됐습니다.
이때 어떤 Dao 클래스에서 책임을 가져야 될지 고민이 됐습니다.
1. CarDao
2. GameDao
3. 새로운 Dao 클래스 생성

저는 원래 가져오는 정보, 즉, 테이블에 따라 Dao를 2개로 나누었습니다. 하지만, JOIN한 결과를 조회하려고 하니 그 어떤 클래스에 두기에도 애매하다는 생각을 했습니다. 그렇다고 새로운 Dao 클래스를 생성하기엔 `조인할 때마다 Dao 클래스를 생성하면 관리 대상이 너무 많아지지 않을까?`라는 생각이 들어, 고민이 됐습니다.

결국 Game 전체에 대한 이력이니, GameDao가 해당 책임을 갖도록 결정했습니다. 완태는 어떤 근거로 선택을 하셨을지 궁금합니다!

wannte replied:
Dao 의 개념을 테이블과 1대1 매칭된다면, JOIN 을 하는 경우 어디에 위치시킬지 까다로워지겠네요.

제 생각은, 지금의  Game 과 Car 의 관계는 Game 이 Car를 가지고 있기 때문에, GameDao 에서 처리하지 않을까 합니다.
또 다른 방법으로는, JOIN 문 자체를 쓰지 않고 어플리케이션 레벨에서 조회 후 aggregation 할 수도 있을 것 같아요.

yujamint replied:
의견 감사합니다! 현재는 DB 구조를 바꾸며 조인을 하지 않고, CAR_RESULT 테이블만 조회하고 있습니다. 
다음에 JOIN을 통해 조회할 일이 있을 때 완태의 의견대로 진행해보겠습니다 :)

yujamint commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -1,37 +1,51 @@
-    public ResponseEntity<ResponseDto> playGame(@RequestBody RequestDto requestDto) {
-        List<String> carNames = NameFormatConverter.splitNameByDelimiter(requestDto.getNames());
+    public ResponseEntity<GamePlayResponseDto> playGame(@RequestBody RequestDto requestDto) {
+        final Names carNames = new Names(requestDto.getNames());
```
### 사용자의 입력값에 대한 유효성 검사는 어떤 계층에서 이뤄져야 하나요?
저는 현재 유효성 검사를 다음과 같이 진행하고 있습니다.
1. 도메인 객체 생성자에서 검사
2. 도메인 객체가 존재하지 않는 값에 대해서는 컨트롤러 메서드 내에서 검사

도메인 객체를 생성하는 책임 역시 컨트롤러가 갖고 있으므로, 컨트롤러에서 모든 유효성 검사를 하고 있습니다.

그리고, 예외 처리하고 있는 내용은 다음과 같습니다.
- 중복된 이름 입력
- 0 이하의 시도횟수 입력
- 1~5자를 벗어나는 길이의 이름 입력

관점에 따라 다르겠지만, 저는 해당 내용들이 어느정도 비즈니스와 관련된 예외 상황이 생각합니다. 그리고 레이어드 아키텍처에선 비즈니스 계층이 비즈니스와 관련된 Validation의 책임을 갖는다고 알고 있습니다.

그렇다면, 제가 컨트롤러에서 검증한 내용들은 모두 서비스 계층에서 검증이 이루어져야 했던 걸까요? 아직 레이어드 아키텍처가 익숙하지 않아서 고민이 되는 것 같습니다.

추가로, `숫자를 입력해야 하는 칸에 문자를 입력한다.`와 같은 입력에 대한 예외 상황 역시 서버가 검증을 해야 되나요? 혹은 뷰에서 필터링된 값이 들어올 것이라고 가정하고 검증하지 않나요?? 후자의 경우 안정성이 떨어질 것 같아 전자를 선택하는 게 안전해 보이긴 합니다. 현업에선 어떤 식으로 진행되는지 궁금합니다.

wannte replied:
각 계층별로 담당/책임의 validation 이 다르다 라고 말할 수 있을 것 같아요.

"중복된 이름을 허용하지 않는다" 라는 규칙이 도메인의 규칙이기에 도메인 단에서 처리할 수 있을 것 같고,
요청의 type/형태 등의 책임은 view 가 가지도록 할 수 있을 것 같아요.

1단계에서 inputView의 책임과 도메인의 책임에 대해서 같은 맥락으로 고민해보셨으면 해요!

yujamint commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -6,33 +6,33 @@ public class Car implements Comparable<Car> {
+        this.position = new Position(CAR_INITIAL_LENGTH);
     }
 
     public void moveByNumber(int moveForNum) {
```
### 레이어드 아키텍처에서, 비즈니스 로직을 갖는 도메인 객체가 존재하는 게 일반적인가요?
현재 `차 이동`, `유효성 검사` 등의 비즈니스 로직은 순수 자바로 작성한 도메인 객체들이 가지고 있습니다.
레이어드 아키텍처에서는 보통 도메인 객체들이 이러한 로직을 갖지 않고, 서비스 클래스가 모든 비즈니스 로직을 갖나요??
만약 서비스 클래스가 모든 로직을 갖는다면, 현재의 도메인 객체가 필요 없어지는 건가요??

1단게에서 작성했던 코드를 통해 웹으로 전환하다보니 도메인 객체들이 비즈니스 로직을 모두 갖고 있고, 서비스 클래스는 이를 호출해서 조립해주는 역할을 하고 있는 것 같습니다. 일반적으로는 어떤 구조를 갖는지 궁금합니다!

wannte replied:
이 부분에 대해서는 의견이 다양한 데, 저는 Service Layer를 가능한 Thin 하게 유지하고 Domain 객체가 로직을 처리하도록 합니다.

관련해서 읽어보면 좋을 문서 첨부드려요!
https://martinfowler.com/bliki/AnemicDomainModel.html

yujamint replied:
완태의 의견 남겨주시며 좋은 글까지 추천해주셔서 감사합니다!! 핵심 비즈니스 로직은 도메인 계층이 갖도록 하고, 서비스 계층에서 이를 호출하며 위임하는 것이 좋다는 의견으로 이해했습니다!!

그렇다면 ORM 기술을 사용하며 테이블과 매핑하는 엔티티 클래스를 생성하게 될 때는 어떤 구조로 설계하는지 궁금합니다. 엔티티 클래스가 비즈니스 로직을 갖게 되는 건가요?? 혹은 POJO 도메인 모델 객체를 하나 더 만드는 건가요?? 즉, 엔티티 클래스 따로, 비즈니스 로직 갖는 도메인 클래스 따로 구현하는지 궁금합니다!! 이에 대해 궁금해하기엔 시기상조일 수 있지만, 궁금함이 생겨 질문드립니다 ㅎㅎ..

+) 완태는 이런 해외 아티클 읽을 때 원문으로 보시나요?! 저는 번역 없이는 보기가 힘드네요 ㅋㅋ...

wannte replied:
ORM 기술에 대해서 현업에 얼마 안되었지만,, 제 경험을 조금 공유해드리면,,

ORM(저 같은 경우에는 JPA) 기술의 발달이 있어, 제 생각으론 충분히 엔티티를 도메인 모델로 사용할 수도 있다 생각했습니다. 하지만, 대부분의 현업자들과 대화를 해봤을 때에는 엔티티는 DB 데이터이기에 로직이 들어가면 안 된다고 생각하시는 분들도 많더라고요. 또한, 엔티티를 도메인 객체로 직접 사용하게 되었을 때의 한계점도 분명히 존재합니다.

 그래서 저는 민트가 말하는 대로 주로 POJO 도메인 모델 객체를 설계하고 여기다 지금까지 우리가 해왔던 validation 과 로직에 대한 책임을 부여하는 편입니다.

해외 아티클은 가능하면 원문으로 읽으려 노력합니다만,, 요즘은 기술의 발전으로 [deepl](https://www.deepl.com/translator) 이라는 플러그인을 활용하여 도움을 받습니다 :) 플러그인을 사용하면 드래그하고 클릭시 바로 번역이 가능합니다

yujamint replied:
그렇군요!! 완태의 경험 공유해주셔서 감사합니다 ㅎㅎ 그렇다면 엔티티 클래스와 도메인 모델 객체를 따로 두기 때문에 각각 엔티티로/도메인 객체로 변환해주는 불편함이 따르겠다는 생각이 드네요. 
이에 대해서는 점차 배워가며 직접 느껴보겠습니다!!

알려주신 덕분에 바로 deepl 플러그인 깔았습니다. 감사합니다!!!!!

wannte commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -1,37 +1,51 @@
+import racingcar.view.OutputView;
 
-@Controller
+@RestController
 public class WebController {
```
동일한 자동차 경주 게임을 Web과 Console 각각에서 진행할 수 있도록, ConsoleController 도 추가 요청드립니다.

잘 만들어진 도메인 구조라면, Console 과 Web에서 큰 변화 없이도 동일하게 동작할 것으로 보입니다.

yujamint replied:
ConsoleController 추가했습니다!! 처음 구현할 당시에는 요구사항을 제대로 이해하지 못했네요. 감사합니다!

wannte commented on src/main/java/racingcar/dao/InMemoryGameDao.java:
```diff
@@ -31,4 +36,36 @@ public int save(int trialCount, String winners) {
+        return new ArrayList<>(historyByGame.values());
+    };
+
+    private GamePlayResponseDto getGameResult(final Map<Integer, GamePlayResponseDto> resultByGameId,
```
GamePlayResponseDto 를 dao 레벨에서 리턴하는 데, 이는 db 로직이 view 로직을 알고 있는 것으로 보입니다.

어떤 장단점이 있을까요?

다양한 결과 응답에 대한 요구사항이 생겼을 때를 고민해보시면 좋을 것 같아요.

yujamint replied:
뷰 - 컨트롤러 계층 간에 데이터를 전달하는 `GamePlayResponseDto`를 dao 레벨에서도 리턴하지 않도록 수정했습니다.
기존의 방식은 변경된 요구사항에 따라 뷰와 Dto를 수정하게 되면 DB 로직까지 수정돼야하는 된다는 문제점을 인지했습니다. 감사합니다!

wannte commented on src/main/java/racingcar/exception/ApiExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(exceptionResponse);
+    }
+}
```
IllegalArgumentException 외의 에러에 대해서도 (Exception.class) 에 대해서도 정의된 에러를 내려주면 좋을 것 같습니다.

그 때에는 어떤 에러 코드를 사용하면 좋을 지 고민해봤으면 해요

wannte commented on README.md:
```diff
@@ -19,3 +19,48 @@ Dto
+- [ ] 사용자의 입력값에 대한 유효성 검사는 어떤 계층에서 이뤄져야 할까?
+- [ ] 현재는 순수 자바 코드로 작성된 도메인 객체가 있고, 해당 도메인에서 유효성 검사가 이뤄지고 있다. 보통은 어떻게 유효성 검사를 할까?
+- [ ] 엔티티는 비즈니스 로직을 가져도 될까?
+- [ ] DB에 종속적이지 않은 테스트를 작성하는 방법
```
>  엔티티는 비즈니스 로직을 가져도 될까?

현재 코드 상에서 민트가 말하는 엔티티는 어떤 것이라 생각하시나요?

 > DB에 종속적이지 않은 테스트를 작성하는 방법

레벨 1에서 순수 자바 객체에서 진행하던 테스트 코드들이 DB가 침투하게 되었을 때, 망가지게 되는데 이걸 어떻게 유지할 수 있을 지 고민해보면 좋을 것 같아요. DB에 대한 '저장/조회'에 대한 테스트라면 DB에 종속적이겠지만, DB에서 조회한 로직을 검증하고 싶다면 DB에 의존하지 않을 수도 있을 것 같네요.

yujamint replied:
> 현재 코드 상에서 민트가 말하는 엔티티는 어떤 것이라 생각하시나요?

제가 알고 있는 엔티티는 테이블과 완전히 매핑되는 클래스입니다. 기존에는 엔티티의 필요성을 느끼지 못해 제 코드에는 사용하지 않았습니다. 하지만 현재는 콘솔 애플리케이션을 구현하며, 순수 자바 코드로 DB에 저장되는 정보를 표현하는 엔티티 클래스를 생성했습니다.
지금도 엔티티를 완전한 필요에 의해, 적절하게 사용하고 있지는 않은 것 같습니다. 오히려 Dto에 가까운 느낌으로 사용하고 있는 것 같습니다. 엔티티의 필요성은 ORM 프레임워크를 사용하며 느끼게 될까요??

테스트에 대한 의견 남겨주셔서 감사합니다. 이번 미션에서는 시간이 부족하여 테스트에 대해서 깊게 고민하지 않았지만, 다음 미션에서는 완태가 남겨주신 의견 참고하며 좋은 테스트를 작성하도록 노력하겠습니다!!!

wannte reviewed (changes requested):
안녕하세요 민트!

빠르게 구현해주셨네요, 리뷰가 조금 늦었습니다.

코멘트 남겼으니 확인부탁드려요~ 🙏

yujamint committed:
refactor: 테이블 스키마 수정과 그에 따른 Dao 수정

데이터의 중복을 없애기 위해 PLAY_RESULT의 winners 컬럼을 제거하고 CAR_RESULT가 is_winner 컬럼을 갖도록 수정하였다.

yujamint committed:
refactor: NamesFormatConverter 책임 Names에게 위임

더이상 컨트롤러와 서비스가 사용하지 않게 되면서 Names가 변환 로직을 갖도록 수정

yujamint committed:
refactor: TryCount 값 포장 객체 생성

yujamint committed:
feat: 콘솔 애플리케이션 구현

yujamint committed:
refactor: dao 반환 타입 수정

dao가 계층에 맞는 dto를 반환하도록 수정

yujamint committed:
feat: Exception.class에 대한 핸들러 추가

yujamint committed:
refactor: 자동차 경주 정보 BatchUpdate를 통해 저장하도록 수정

yujamint committed:
refactor: dao 패키지 분리

yujamint committed:
refactor: DB를 연결하지 않는 Cache Dao 구현

yujamint committed:
refactor: 콘솔 애플리케이션은 스프링부트를 통해 실행되지 않도록 수정

yujamint committed:
feat: 콘솔 애플리케이션 구현

yujamint committed:
feat: Exception Handler 로깅 추가

yujamint committed:
docs: 2단계 기능 완료 처리

yujamint commented on src/main/java/racingcar/exception/ApiExceptionHandler.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<ExceptionResponse> handleUnCaughtException(Exception e) {
```
`Exception.class`로 catch 되는, 즉, 정의되지 않은 예외에 대해서는 서버의 문제로 간주하고 상태 코드로 500(INTERNAL_SERVER_ERROR)을 사용하는 것으로 결정했습니다. 발생할 수 있는 예외를 처리하지 않은 서버의 문제라고 생각했기 때문입니다.

추가로, 정의되지 않은 예외가 발생했다는 것은 개발자가 빠르게 인지하고, 개선해야 되는 문제라고 생각하여 로그를 출력하도록 했습니다. 아직 로깅에 대해서 깊은 이해를 갖고 있지 않아서 올바른 방향으로 사용한 것인지 확신은 없네요..!

각 상황에 맞는 예외 메세지를 응답하고, 로그를 남기는 것이 최선이겠지만, `Exception.class`와 같이 구체적으로 어떤 예외 상황인지 모르는 상황에서는 어떻게 처리하는 것이 최선일지 궁금합니다!

wannte replied:
넵, 잘 적용해주신 것 같습니다.
다만, 현재 로그로는 어떤 Exceptino이 발생한 것이 확인이 불가능해서, 어떤 에러인지 남길 수 있다면, 좋을 것 같습니다.

추가로, 고객한테 이 메시지가 전달되도 될지에 대해서 고민해 보시면 좋을 것 같아요.

yujamint replied:
로그를 통해 어떤 Exception이 발생했는지 알 수 있도록 수정했습니다!
추가로, Exception.getMessage()가 아닌 메세지를 전달하도록 수정했습니다.

고객에게 예외 메세지가 전달되면 안 되는 이유에 대해서 생각을 해보며 한 가지 궁금한 점이 생겼습니다!
예외 메세지는 개발자가 보게 될 것인데, 이를 고객에게 전달하면 의도치 않은 정보까지 전달할 수 있고, 이는 곧 보안의 문제로 이어지기 때문인 걸까요?
그렇다면, 저는 기존에 클라이언트(개발자)를 대상으로 정의한 메세지를 최종 사용자(고객)에게까지 전달하려고 했었기에 문제의 여지가 있었던 건가요?!

추가로, 제가 생각하는 **클라이언트(개발자)를 대상으로 정의한 메세지**와 **최종 사용자를 대상으로 정의한 메세지**의 차이점을 남기겠습니다.
- 클라이언트 대상: 해당 예외를 복구할 수 있도록, 관련된 많은 정보를 담은 메세지 전달
- 최종 사용자 대상: `비밀번호는 영어 대문자와 특수문자를 포함해야 합니다.`와 같은 사용자 경험을 중점적으로 고려한 안내 매세지 전달

wannte replied:
클라이언트 대상이더라도, 우린 개발자 도구를 통해 확인할 수 있기 때문에 위험하다 생각합니다.

사용자에게 전달되야 하는 에러에 대해서는 다른 방식으로 약속?정의하고 보여줄 수 있도록 해야 할 것 같아요

yujamint commented on src/main/java/racingcar/controller/dto/GamePlayResponseDto.java:
```diff
@@ -0,0 +1,47 @@
+        this.racingCars = convert(racingCars);
+    }
+
+    public GamePlayResponseDto(List<GameFinishedCarDto> cars) {
```
View - Controller - Service 계층이 주고받는 Dto인 `GamePlayResponseDto`의 생성자로 Service - Repository 계층이 주고받는 Dto인 `GameFinishedCarDto` 객체를 넘겨줘도 괜찮을까요?
즉, A-B 계층에서 사용하는 Dto 객체를 C-D 계층에서 사용되는 Dto의 생성자로 전달해도 괜찮을까요??

서비스에서 `GameFinishedCarDto`의 값을 까서 `GamePlayResponseDto`를 만들어 주기에는 서비스의 코드가 지저분해지는 것 같고, 반대로 `GameFinishedCarDto`를 `GamePlayResponseDto` 생성자에 전달하는 것은 Dto 생성자 내에 데이터를 변환하는 로직이 생기는 것 때문에 의문이 생기는 것 같습니다.
데이터 변환은 비즈니스 로직이 아니니 상관 없을까요??..

wannte replied:
이 부분에 대한 생각도 다양할 것 같은데,

쿼리 로직에 대해서는 중간 도메인 객체를 거치지 않고 결과 값을 반환해줄 수도 있다 생각합니다.

---

그 이전에 저는 dto  반대파 이기 때문에,,, :sob:

yujamint replied:
> 쿼리 로직에 대해서는 중간 도메인 객체를 거치지 않고 결과 값을 반환해줄 수도 있다 생각합니다.

이 말에 대해서 제대로 이해하지 못 했습니다..! Dao 레벨에서는 Dto가 아닌, 도메인 객체/엔티티를 그대로 서비스에 반환해도 된다는 말씀이신 걸까요?!

wannte replied:
기존의 서비스 로직은

dao/dto -> domain -> service/dto 의 순서로 진행되고 잇습니다.
불필요하다면, 중간 doamin 변환 로직은 진행하지 않아도 된다 생각합니다.

---

별개로, 의존 방향을 해치지 않는다면,  `도메인 객체/엔티티를 그대로 서비스에 반환해도 된다`라고 생각합니다.

yujamint commented on src/main/java/racingcar/dao/car/InMemoryCarDao.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    public void saveAll(int gameId, List<Car> cars, List<String> winners) {
```
`batchUpdate`를 사용하여 여러 건의 데이터를 insert할 때 성능을 개선할 수 있다는 것을 알게 되었습니다. 그래서 학습을 목적으로 `batchUpdate`를 사용했습니다. 

그리고 `System.currentTimeMillis`를 통해 시간을 비교해봤습니다. 10개 정도의 데이터를 저장할 땐 반복문으로 쿼리를 날리는 것과 `batchUpdate`의 성능 차이가 크지 않는 것을 보고 궁금증이 생겼습니다.

현업에서는 어떤 기준으로 성능 개선을 고려하기 시작하나요?? 대부분의 로직을 실행하며 시간 측정을 하고, 그 과정에서 고려를 하게 되나요? 혹은 100번, 200번의 쿼리를 날려야 되는 상황처럼 직관적으로 성능 저하가 의심될 때 고려를 하게 되나요?
당장 제가 직면한 문제는 아니지만, 현업에 계신 완태의 의견이 궁금하여 질문 남깁니다!!

wannte replied:
개인적인 생각 말씀 드리면, 조직의 특성에 따라 달라진다 생각합니다.

제가 속한 조직은 개발에 빠른 속도를 중요시여기기 때문에, 이러한 (과도한 상황이 예측되지 않는 다면) 성능 적인 고민을 먼저 하기보다는 일이 되게 함을 중요시 여깁니다. 이후에 API latency나 DB Slow Query 를 통해 탐지하고 해결하는 정도로 진행한다고 보시면 될 것 같습니다.

여기서 테스트 속도가 차이가 없는 것에 대해 추가 코멘트 드리면,
지금의 조회는 내 local pc에서 진행되기 때문에, 그 속도의 차이가 크지 않습니다. 만약 DB가 다른 곳 (ex. 클라우드 환경)에 있다면, 네트워크 latency로 인해 더 큰 차이가 발생하지 않을까 생각합니다.(이후 경험하겠지만, 네트워크 비용이 제일 비쌉니다!)

yujamint replied:
역시나 조직에 따라 다르겠군요..! 완태의 경험 말씀해 주셔서 감사합니다!! 얼른 성장해서 현업을 경험해보고 싶네요 ㅎㅎ

저는 속도의 차이가 별로 없는 점에 대해서, 실행되는 쿼리 수가 적은 케이스로 테스트했기 때문이라고 생각했습니다. 네트워크 비용은 전혀 생각 안 하고 있었네요. 코멘트 남겨주셔서 감사합니다!

yujamint reviewed (commented):
안녕하세요 완태! 리뷰이 민트입니다 :)
2단계 리뷰 내용에 따라 리팩터링 진행하여 다시 리뷰 요청드립니다!
이번에 수정된 내용은 크게 다음과 같습니다.

- dao 개선 
  - dao 레벨에서 반환하는 dto를 새로 생성함으로써 view에 전달되는 dto를 반환하지 않도록 수정했습니다.
  - 학습 목적으로 batchUpdate를 사용하며 Insert 문을 처리했습니다.
- 콘솔 애플리케이션 구현
  - DB에 접근하지 않는 Dao를 구현했습니다. 이 과정에서 Entity 클래스를 구현했습니다.
- Exception.class 핸들러 추가
  - 이에 대한 추가적인 대처(?)로 로깅을 사용해봤습니다.

이번 리뷰도 잘 부탁드리겠습니다 완태!! 감사합니다 ㅎㅎ

wannte commented on src/main/java/racingcar/dao/car/CacheCarDao.java:
```diff
@@ -0,0 +1,38 @@
+import racingcar.dao.dto.GameFinishedCarDto;
+import racingcar.model.Car;
+
+public class CacheCarDao implements CarDao {
```
Cache라는 명명 보다는 InMemory 정도가 어떨까요?
Cache라는 개념은 다양하게 사용될 수 있을 것 같아, 지금의 사용을 유추하기 어려울 것 같습니다.

wannte commented on src/main/java/racingcar/dao/car/InMemoryCarDao.java:
```diff
@@ -0,0 +1,62 @@
+import racingcar.model.Car;
+
+@Repository
+public class InMemoryCarDao implements CarDao {
```
앗. 여기서 InmemoryCarDao 가 있었네요

오히려 이 부분은 JdbcTemplate 을 사용하기 때문에 DB 스러운 이름으로 바꿀 수 있을 것 같습니다.

wannte commented on src/main/java/racingcar/service/GameService.java:
```diff
@@ -22,13 +27,23 @@ public GameService(NumberGenerator numberGenerator, CarDao carDao, GameDao gameD
 
-    public void executeRacingGame(Cars cars, int tryCount) {
-        for (int count = 0; count < tryCount; count++) {
+    public void executeRacingGame(Cars cars, TryCount tryCount) {
+        for (int count = 0; count < tryCount.getValue(); count++) {
             cars.moveResult(numberGenerator);
         }
+        int gameId = gameDao.save(tryCount.getValue());
+        carDao.saveAll(gameId, cars.getCars(), cars.getWinners());
+    }
+
```
과할 수는 있지만, car.moveResult 를 진행하는 Game 객체가 있다면, 로직이 더 깔끔해질 수 있을 것 같아요.

wannte reviewed (changes requested):
안녕하세요, 민트!

여러 질문 남겨주셨는데, 관련해서 추가 코멘트 남겼습니다. 확인 부탁드려요!

yujamint committed:
refactor: Dao 클래스명 변경

- InMemory -> JdbcTemplate
- Cache -> InMemory

yujamint committed:
refactor: exception handler 개선

- 어떤 예외가 발생했는지 로그 출력
- 예외 메세지 수정

yujamint committed:
refactor: Game 도메인 객체 생성

시도횟수만큼 게임을 진행하는 책임 수행

yujamint reviewed (commented):
안녕하세요 완태! 리뷰이 민트입니다.
리뷰 주신 내용 반영 완료하여 리뷰 요청드립니다!
궁금한 내용 코멘트에 남겼으니 확인해주시면 감사하겠습니다 :)
감사합니다!!

wannte commented on src/main/java/racingcar/exception/ApiExceptionHandler.java:
```diff
@@ -21,9 +21,9 @@ public ResponseEntity<ExceptionResponse> handleIllegalArgumentException(IllegalA
-        logger.error("정의하지 않은 예외가 발생했습니다.");
-        final ExceptionResponse exceptionResponse = new ExceptionResponse(e.getMessage());
-        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
+        logger.error("정의하지 않은 예외 발생: " + e.getClass());
```
`log.error` 정의 된 부분을 확인해보면, throwable 을 입력 받는 시그니처가 있는데, 이 부분을 확인해보셔도 좋을 것 같습니다.

wannte reviewed (approved):
안녕하세요, 민트!

빠르게 리뷰 만영해주셨네요! 요구사항은 충분히 만족한 것 같아 이만 머지하곘습니다! 🙇
